modules/launcher/src/main/java/org/mule/runtime/module/launcher/plugin/DefaultApplicationPluginRepository.java

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright (c) MuleSoft, Inc.  All rights reserved.  http://www.mulesoft.com
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+
+package org.mule.runtime.module.launcher.plugin;
+
+import static java.io.File.separator;
+import static java.util.Collections.EMPTY_LIST;
+import static java.util.Collections.unmodifiableList;
+import static org.apache.commons.io.FileUtils.forceDelete;
+import static org.apache.commons.io.FilenameUtils.removeExtension;
+import static org.apache.commons.io.IOCase.INSENSITIVE;
+import static org.mule.runtime.core.util.FileUtils.unzip;
+import static org.mule.runtime.core.util.Preconditions.checkArgument;
+import static org.mule.runtime.module.launcher.MuleFoldersUtil.getContainerAppPluginsFolder;
+
+import java.io.File;
+import java.io.FileFilter;
+import java.io.IOException;
+import java.util.LinkedList;
+import java.util.List;
+
+import org.apache.commons.io.filefilter.DirectoryFileFilter;
+import org.apache.commons.io.filefilter.SuffixFileFilter;
+
+/**
+ * Default implementation of an {@link ApplicationPluginRepository} that holds in memory the list
+ * of artifact plugin descriptors bundled with the container.
+ *
+ * @since 4.0
+ */
+public class DefaultApplicationPluginRepository implements ApplicationPluginRepository
+{
+    private List<ApplicationPluginDescriptor> containerApplicationPluginDescriptors;
+    private final ApplicationPluginDescriptorFactory pluginDescriptorFactory;
+
+    /**
+     * @param pluginDescriptorFactory a {@link ApplicationPluginDescriptorFactory} for creating from the container applications plugins folder the list of {@link ApplicationPluginDescriptor}'s
+     */
+    public DefaultApplicationPluginRepository(ApplicationPluginDescriptorFactory pluginDescriptorFactory)
+    {
+        checkArgument(pluginDescriptorFactory != null, "Application plugin descriptor factory cannot be null");
+        this.pluginDescriptorFactory = pluginDescriptorFactory;
+    }
+
+    @Override
+    public synchronized List<ApplicationPluginDescriptor> getContainerApplicationPluginDescriptors() throws IOException
```

Comment:
Do not synchronize the method, synchronize only the containerApplicationPluginDescriptors initialization part